engine/crates/tracing/src/otel/metrics.rs

```diff
@@ -0,0 +1,90 @@
+use opentelemetry::{KeyValue, Value};
+use opentelemetry_sdk::metrics::data::Temporality;
+use opentelemetry_sdk::metrics::reader::{AggregationSelector, TemporalitySelector};
+use opentelemetry_sdk::metrics::{Aggregation, InstrumentKind, PeriodicReader, SdkMeterProvider};
+use opentelemetry_sdk::Resource;
+use std::time::Duration;
+
+use crate::config::TracingConfig;
+use crate::error::TracingError;
+
+/// Sets up the metrics provider
+pub fn setup_metrics<'a>(
+    config: &TracingConfig,
+    resource_attributes: impl IntoIterator<Item = (&'a str, Value)>,
+) -> Result<(), TracingError> {
+    setup_meter_provider(
+        config,
+        resource_attributes
+            .into_iter()
+            .map(|(k, v)| KeyValue::new(k.to_string(), v))
+            .collect(),
+    )
+}
+
+fn setup_meter_provider(config: &TracingConfig, resource_attributes: Vec<KeyValue>) -> Result<(), TracingError> {
+    let runtime = opentelemetry_sdk::runtime::Tokio;
+
+    let mut provider = SdkMeterProvider::builder().with_resource(Resource::new(resource_attributes));
+
+    if config
+        .exporters
+        .stdout
+        .as_ref()
+        .map(|cfg| cfg.enabled)
+        .unwrap_or_default()
+    {
+        let reader = PeriodicReader::builder(opentelemetry_stdout::MetricsExporter::default(), runtime.clone())
+            .with_interval(Duration::from_secs(10))
+            .with_timeout(Duration::from_secs(10))
+            .build();
+
+        provider = provider.with_reader(reader);
+    }
+
+    #[cfg(feature = "otlp")]
+    if let Some(config) = config.exporters.otlp.as_ref().filter(|cfg| cfg.enabled) {
+        use opentelemetry_otlp::MetricsExporterBuilder;
+        let exporter = crate::otel::provider::build_otlp_exporter::<MetricsExporterBuilder>(config)?
+            .build_metrics_exporter(Box::new(DeltaTemporality), Box::new(AggForLatencyHistogram))
+            .map_err(|e| TracingError::MetricsExporterSetup(e.to_string()))?;
+        let reader = PeriodicReader::builder(exporter, runtime.clone())
+            .with_interval(Duration::from_secs(10))
+            .with_timeout(Duration::from_secs(10))
+            .build();
+
+        provider = provider.with_reader(reader);
+    }
+
+    opentelemetry::global::set_meter_provider(provider.build());
```

Comment:
we shouldn't be doing this here. We should return the provider back to the user of this crate, and that's where it should decided if its global or not.